lib/vdsm/virt/vm.py

```diff
@@ -2813,32 +2814,44 @@ def _run(self):
             hooks.before_vm_start(self._buildDomainXML(), self._custom)
 
             fromSnapshot = self._altered_state.from_snapshot
-            srcDomXML = self._src_domain_xml
-            if fromSnapshot:
-                # If XML was provided by Engine, disk paths have already been
-                # corrected in __init__.  If legacy configuration
-                # (e.g. 'vmName') is present, we use the legacy path here.
-                # Otherwise we leave srcDomXML untouched, since we don't have
-                # anything from Engine (4.2.0) to update it with.
-                if 'vmName' in self.conf:
-                    srcDomXML = self._correct_disk_volumes_from_conf(srcDomXML)
-                srcDomXML = self._correctGraphicsConfiguration(srcDomXML)
-            hooks.before_vm_dehibernate(srcDomXML, self._custom,
-                                        {'FROM_SNAPSHOT': str(fromSnapshot)})
-
-            # TODO: this is debug information. For 3.6.x we still need to
-            # see the XML even with 'info' as default level.
-            self.log.info("%s", srcDomXML)
-
-            self._connection.defineXML(srcDomXML)
             restore_path = self._altered_state.path
             fname = self.cif.prepareVolumePath(restore_path)
             try:
                 if fromSnapshot:
-                    self._connection.restoreFlags(
-                        fname, srcDomXML, libvirt.VIR_DOMAIN_SAVE_PAUSED)
+                    srcDomXML = self._src_domain_xml
+                    # If XML was provided by Engine, disk paths have already
+                    # been corrected in __init__.  If legacy configuration
+                    # (e.g. 'vmName') is present, we use the legacy path here.
+                    # Otherwise we leave srcDomXML untouched, since we don't
+                    # have anything from Engine (4.2.0) to update it with.
+                    if 'vmName' in self.conf:
+                        srcDomXML = self._correct_disk_volumes_from_conf(
+                            srcDomXML)
+                    srcDomXML = self._correctGraphicsConfiguration(srcDomXML)
                 else:
-                    self._connection.restore(fname)
+                    srcDomXML = self._connection.saveImageGetXMLDesc(fname)
```

Comment:
This deserves commentary in the commit message -- we use a different srcDomXML now.  Does it matter?  Can there be changes performed in `Vm.__init__` on it that could be lost here?

Reply:
You are right, this indeed deserves a comment. I have updated the commit message. For visibility here it is too:

    Domain XML that is used to restore the VM is loaded from libvirt's
    hibernation image which replaces XML passed from API call and that goes
    through init. This should not lead to any change in behavior. While we
    used the passed XML to create the domain it has been always immediately
    replaced by libvirt in the resume call. That means if we are doing any
    changes to the XML during early init such changes were always lost and
    will still remain lost.

Reply:
So this means:

- The hook obtains a more correct XML.
- The content of the domain XML passed to defineXML doesn't matter.
- libvirt takes and updates CPU pinning from the second argument of restoreFlags call.

Right?  Did you verify the last two facts above (libvirt documentation isn't very clear in those areas)?

Reply:
> So this means:
> 
>     * The hook obtains a more correct XML.

Yes

> 
>     * The content of the domain XML passed to defineXML doesn't matter.

Yes, that is my suspicion. I planned to investigate it little bit more but right know it seems to me the `defineXML` is superfluous and has no effect. The complete domain XML is then replaced with the XML libvirt has (or with what we pass in `restoreFlags`).

> 
>     * libvirt takes and updates CPU pinning from the second argument of restoreFlags call.

Yes. 

> 
> Right? Did you verify the last two facts above (libvirt documentation isn't very clear in those areas)?

The second one yes, the first one I still want to investigate more. But this probably does not need to block the PR.

It occurred to me I have omitted another vital piece of information so I have updated the commit message one more time. Again for visibility:

    It may be tempting to pass domain XML from engine or from our pickled data
    to `resumeFlags()`. Neither is possible because our domain XML is not
    migration XML. Doing so would end in error because libvirt would fail to
    match the guest specific portion with the stored domain XML.

Reply:
OK.  Please check whether there is any relation between the defineXML domain XML and restoreFlags ones (probably the best way is to ask on the libvirt users list).  But yes, we needn't block this patch on this.